### DISCUSSION TITLE
Can't get postcodes to work

### DISCUSSION THREAD

**Discussion by mausch** [osm-search/Nominatim#2651] [Q&A] [answered]
Hi! I imported GB+Ireland (britain-and-ireland-latest.osm.pbf from https://download.geofabrik.de/europe.html)
I made a mistake and got the old SQL postcodes instead of the new CSV file.

So to get postcodes imported I followed the instructions in https://nominatim.org/release-docs/latest/admin/Import/#external-postcodes and https://nominatim.org/release-docs/latest/customize/Postcodes/  and I ran in the project dir:

```
$ wget https://www.nominatim.org/data/gb_postcodes.csv.gz
$ nominatim refresh --postcodes
2022-03-29 13:46:28: Using project directory: /nominatim
2022-03-29 13:46:28: Update postcodes centroid
2022-03-29 13:46:31: Starting indexing postcodes using 1 threads
2022-03-29 13:46:31: Starting postcodes (location_postcode) (using batch size 20)
2022-03-29 13:46:31: Done 0/0 in 0 @ 0.000 per second - FINISHED postcodes (location_postcode)
```

i.e. I got no postcodes updated? My `location_postcode` table remains empty.

I looked around the code and ran [this query](https://github.com/osm-search/Nominatim/blob/c3788d765ed4e7ddf00794085af757714bc102cf/nominatim/tools/postcodes.py#L172-L184) and got no results.

Simplifying that a bit, I get no results for:
```
SELECT *
FROM place
WHERE address ? 'postcode'
```

I do have 8728 rows in `place` with non-null address (out of a total 2.8M rows in `place`)

Maybe this pbf file I imported is missing some necessary data?
Any idea where things are going wrong here?

Many thanks.

**Comment by mtmail** [comment, marked as answer]
I tested a similar scenario with a planet install today. Removed the file from the project directory, ran `nominatim refresh --postcodes`, then added the file again, ran `nominatim refresh --postcodes`. Difference of about 1.4 million entries in the `location_postcode` table (`where country_code='gb'`).

Your output indicates the `Update postcodes centroid` took only 3 seconds. That would be super fast for a full table scan of the placex table, since the query doesn't use any indices.

Did the import finish? Does `nominatim admin --check-database` report any issues?

On a planet database my counts are
```sql
select count(*) from placex where country_code='gb';
 7885047
select count(distinct(address->'postcode')) from placex where country_code='gb' and address->'postcode' is not null;
 278152
```

(The import and replication fill the `place` table, then index the data into the `placex` table. For majority of debugging purposes you can ignore the `place` table).

**Comment by mausch** [comment]
I built a new instance of Nominatim and this time I got the correct postcodes file in place when I ran `nominatim add-data`...
but I'm still not getting any postcodes.

No issues reported by `nominatim admin --check-database`

`nominatim refresh --postcodes` still doesn't seem to do anything.

The `location_postcode` table is still empty.

These are the counts I get for those queries:

```sql
select count(*) from placex where country_code='gb';
-- 2365336
select count(distinct(address->'postcode')) from placex where country_code='gb' and address->'postcode' is not null;
-- 0
SELECT count(*) FROM place WHERE address ? 'postcode'
-- 0
```

BTW this is with Nominatim 4.0.1

**Reply by mtmail** [reply]
So 5.2 million less entries in placex. Which import style did you use? https://nominatim.org/release-docs/4.0.1/admin/Import/#filtering-imported-data (default is "extratags"). I think with "street" you would've not imported any addresses. It would still be surprising why `nominatim refresh --postcodes` behaves like this but then the issue is narrowed-down.

```sql
 select rank_address, count(*) from placex where country_code='gb' group by rank_address order by rank_address;
 rank_address |  count
--------------+---------
            0 |  261010
            4 |       1
            5 |      15
            8 |       4
           10 |      16
           12 |     227
           14 |       2
           16 |   11239
           18 |    7204
           20 |   32682
           22 |  111489
           24 |       2
           25 |   39152
           26 | 2063535
           27 |  170971
           30 | 5191043
(16 rows)
```

**Reply by mausch** [reply]
Thank you, that was it! I was importing with 'street'
